### PR TITLE
Bitwise the operations of `OpCode.BOOLAND` and `OpCode.BOOLOR`.

### PR DESCRIPTION
**Pull request by Liaojinghui**
BenchmarkDotNet=v0.13.0, OS=macOS Big Sur 11.4 (20F71) [Darwin 20.5.0]
Intel Core i7-9750H CPU 2.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=5.0.302
  [Host]     : .NET 5.0.8 (5.0.821.31504), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET 5.0.8 (5.0.821.31504), X64 RyuJIT

`true | true`

|                  Method |      Mean |     Error |    StdDev |    Median |
|------------------------ |----------:|----------:|----------:|----------:|
|     Instruction_BOOLAND | 0.0177 ns | 0.0172 ns | 0.0315 ns | 0.0000 ns |
| Instruction_BOOLAND_Bit | 0.0000 ns | 0.0000 ns | 0.0000 ns | 0.0000 ns |
|      Instruction_BOOLOR | 0.2387 ns | 0.0204 ns | 0.0191 ns | 0.2403 ns |
|  Instruction_BOOLOR_Bit | 0.0022 ns | 0.0033 ns | 0.0029 ns | 0.0006 ns |

`false | true`

|                  Method |      Mean |     Error |    StdDev |    Median |
|------------------------ |----------:|----------:|----------:|----------:|
|     Instruction_BOOLAND | 0.2717 ns | 0.0087 ns | 0.0072 ns | 0.2733 ns |
| Instruction_BOOLAND_Bit | 0.0042 ns | 0.0089 ns | 0.0083 ns | 0.0000 ns |
|      Instruction_BOOLOR | 0.0009 ns | 0.0039 ns | 0.0036 ns | 0.0000 ns |
|  Instruction_BOOLOR_Bit | 0.0189 ns | 0.0115 ns | 0.0107 ns | 0.0214 ns |

`true | false`

|                  Method |      Mean |     Error |    StdDev |    Median |
|------------------------ |----------:|----------:|----------:|----------:|
|     Instruction_BOOLAND | 0.0019 ns | 0.0054 ns | 0.0048 ns | 0.0000 ns |
| Instruction_BOOLAND_Bit | 0.0066 ns | 0.0088 ns | 0.0078 ns | 0.0022 ns |
|      Instruction_BOOLOR | 0.2428 ns | 0.0213 ns | 0.0188 ns | 0.2388 ns |
|  Instruction_BOOLOR_Bit | 0.0034 ns | 0.0058 ns | 0.0054 ns | 0.0000 ns |

`false | false`

|                  Method |      Mean |     Error |    StdDev |    Median |
|------------------------ |----------:|----------:|----------:|----------:|
|     Instruction_BOOLAND | 0.2832 ns | 0.0164 ns | 0.0145 ns | 0.2868 ns |
| Instruction_BOOLAND_Bit | 0.0000 ns | 0.0000 ns | 0.0000 ns | 0.0000 ns |
|      Instruction_BOOLOR | 0.0002 ns | 0.0008 ns | 0.0012 ns | 0.0000 ns |
|  Instruction_BOOLOR_Bit | 0.0085 ns | 0.0089 ns | 0.0084 ns | 0.0067 ns |

### CODE REVIEW

File: src/neo-vm/ExecutionEngine.cs [hunk 1]
```diff
@@ -833,6 +833,7 @@ private void ExecuteInstruction()
                     {
                         var x2 = Pop().GetInteger();
                         var x1 = Pop().GetInteger();
+                        if (x2.IsZero) throw new InvalidOperationException($"The value {x2} is out of range.");
```
[the remark this answers is not in the document]
Comment (Liaojinghui) [reply src/neo-vm/ExecutionEngine.cs:836]: Yep, it will throw `DivideByZeroException` automatically.

[the remark this answers is not in the document]
Comment (Liaojinghui) [reply src/neo-vm/ExecutionEngine.cs:836]:
```suggestion
```

File: src/neo-vm/ExecutionEngine.cs [hunk 2]
```diff
@@ -1272,6 +1272,8 @@ private void ExecuteInstruction()
                     {
                         VMArray x = Pop<VMArray>();
                         int index = x.Count - 1;
+                        if (index < 0)
+                            throw new InvalidOperationException($"The value {index} is out of range.");
                         Push(x[index]);
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:1277]: if `index it's < 0` it will throw the exception here, isn't it?

**Comment by shargon** [comment]
Acording to your results, sometimes it's slower

**Comment by Liaojinghui** [comment]
> Acording to your results, sometimes it's slower

Yes, logic operations can have shorter conditions than bitwise conditions under certain cases. No matter, bitwise still outperform the logic operations in general on Bool.

**Comment by erikzhang** [comment]
Do we really need to optimize these few nanosecond operations?

**Comment by Liaojinghui** [comment]
> Do we really need to optimize these few nanosecond operations?

Well, when we break into these opcode, basically all of them are of few nanoseconds. What I put here is a clue on how to optimize the VM, but we still need large scale transaction tests to know how much can it improve the performance.

**Pull request closed without merging**